genai-perf/genai_perf/goodput_calculator/llm_goodput_calculator.py

```diff
@@ -0,0 +1,183 @@
+        """
+        invalid_slos = []
+        self._valid_time_related_slos = {}
+        self._valid_throughput_related_slos = {}
+        for slo_name, slo_value in self._goodput_constraints.items():
+            if slo_name in self._valid_metric_names:
+                if slo_name in self._valid_time_related_names:
+                    self._valid_time_related_slos[slo_name] = (
+                        slo_value * self.MS_TO_NS_CONVERSION
+                    )
+                elif slo_name in self._valid_throughput_related_names:
+                    self._valid_throughput_related_slos[slo_name] = slo_value
+            else:
+                invalid_slos.append(slo_name)
+        if self._valid_time_related_slos:
+            self._has_time_target = True
+        if self._valid_throughput_related_slos:
+            self._has_throughput_target = True
+        if invalid_slos:
+            valid_slos_list = ", ".join(self._valid_metric_names)
+            logger.info(
+                f"Invalid SLOs found: {', '.join(invalid_slos)}. "
+                f"The goodput will be -1. Valid SLOs are: {valid_slos_list}."
+            )
+            self._goodput = self.INVALID_GOODPUT
+
+    def _combine_requests_metric_values(self) -> None:
+        """
+        Combine values from the metrics that match with the valid SLOs at a
+        per request level.
+        """
+        if self.goodput == self.INVALID_GOODPUT:
+            return
+
+        if self._has_time_target:
+            time_names = [
+                self.get_slo_base_name(key) for key in self._valid_time_related_slos
+            ]
+            requests_time_metric_values = [
+                self._metric.data[name] for name in time_names
+            ]
+
+            self._combined_requests_time_metric_values = list(
+                zip(*requests_time_metric_values)
+            )
+
+        if self._has_throughput_target:
+            throughput_names = [
+                self.get_slo_base_name(key)
+                for key in self._valid_throughput_related_slos
+            ]
+            requests_throughput_metric_values = [
+                self._metric.data[name] for name in throughput_names
+            ]
+
+            self._combined_requests_throughput_metric_values = list(
+                zip(*requests_throughput_metric_values)
+            )
+
+    def _count_good_reqs(self) -> Optional[int]:
+        """Count the number of good requests according to SLOs."""
+        if self.goodput == self.INVALID_GOODPUT:
+            return None
+        target_time_metric_values = []
+        target_throughput_metric_values = []
+        if self._has_time_target:
+            num_of_requests = len(self._combined_requests_time_metric_values)
+            target_time_metric_values = list(self._valid_time_related_slos.values())
+        if self._has_throughput_target:
+            num_of_requests = len(self._combined_requests_throughput_metric_values)
```

Comment:
if you have a throughput and time target, num_of_requests will be overwritten here.

Reply:
Yeah, in that case the num_of_requests is still the same number which equals to the total number of requests.

Reply:
that is always valid?

Reply:
A little bit hard to answer this. : )
Yes, because for every completed request, we assume PA populates the information we need in the json and data parser would compute those metrics for the particular request. 
No, if there is something broken with PA. In that case, I believe GenAI-Perf won't work anyways.